Implement JUnit report and XRay integration for CI tests

lucas-p-bernardes opened the pull request:
Gitlab CI pipeline: https://gitlab.com/toradex/rd/torizon-core/torizoncore-builder/-/pipelines/1818264201

lucas-p-bernardes committed:
tests/integration: Implement JUnit test report

This implementation of the JUnit report in this repository is part of a series
of implementations across various Toradex repositories, aiming to standardize
test reporting and facilitate the aggregation of test results for all Torizon
projects.
For this, 'test-torizoncore-common' was split into two executions, one for
Intel and one for Raspberry, so each pipeline runs only one test execution.
In addition to speeding up tests by increasing parallelism, the JUnit report
becomes simpler when only a single BATS test results file is provided.

Related-to: IN-11
Related-to: TCB-506

Signed-off-by: Lucas Bernardes <lucas.bernardes@toradex.com>

lucas-p-bernardes committed:
tests/integration: Implement XRay for jobs running in nightly of TOS7

This change adds support for Xray report uploads to Jira, enabling better
visibility into the current status of TCB with the latest nightly builds of
Torizon. Since the goal is to reflect the integration state with the most
recent system image, the reporting is limited to nightly test jobs only.

Related-to: IN-11

Signed-off-by: Lucas Bernardes <lucas.bernardes@toradex.com>

lucas-p-bernardes committed:
tests/integration: Improve logging of xray-junit-uploader-wrapper

Enhance the xray-junit-uploader-wrapper script by adding informative log
messages that clarify which report is being uploaded, which test execution key
is being used, and which test plan is targeted. This improves visibility during
CI runs.

Related-to: IN-11

Signed-off-by: Lucas Bernardes <lucas.bernardes@toradex.com>

rborn-tx commented on tests/integration/run.sh:
```diff
@@ -223,8 +223,18 @@ echo -e "Starting integration tests...\n"
 if [ "$TCB_REPORT" = "1" ]; then
-    $BATS_CMD 2>&1 | tee $REPORT_FILE
+    $BATS_CMD
+    # ensure the report writing is finished
+    if ! is_report_complete "$REPORT_DIR/report.xml"; then
+        echo "Error: $REPORT_DIR/report.xml not created or incomplete!"
+        exit 6
+    fi
```
@lucasbernardestoradex Can you explain why we need this logic here? If BATS finished executing then the report must have been fully written already, right?

lucas-p-bernardes replied:
@rborn-tx I removed this and executed a pipeline to show what's was happening: https://gitlab.com/toradex/rd/torizon-core/torizoncore-builder/-/pipelines/1827712459/

Note that in https://gitlab.com/toradex/rd/torizon-core/torizoncore-builder/-/pipelines/1827712459/test_report you can see that some reports failed due to "JUnit XML parsing failed". If you open one of this pipelines, and download the XML file in "Job Artifacts" area, you notice that the content of the file is just this:

```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites time="916">
```

Seems the process was finishing before fully write XML file, and returning this XML with missing parts.
This is the problem, but I don't fully understand why the CI was cutting the XML and maybe this solution is not the better one.

rborn-tx replied:
@lucasbernardestoradex I'd suggest further investigating this issue to determine its cause. My concern is that we could be hiding a bigger problem by simply waiting the file to be written since we don't have an explanation for why this wait is needed.

lucas-p-bernardes replied:
Ok, I'm taking a look!

lucas-p-bernardes replied:
@rborn-tx Indeed, there was a bug in bats-core, that was already fixed. So I needed to bump bats-core to catch this fix.
Added commit, that I explain what happened: 
`tests/integration: Bump BATS versions to get fix in --report-formatter junit`

lucas-p-bernardes replied:
Pipeline: https://gitlab.com/toradex/rd/torizon-core/torizoncore-builder/-/pipelines/1860905628/

rborn-tx replied:
Glad to hear.

rborn-tx commented on .gitlab-ci.yml:
```diff
@@ -58,16 +61,16 @@ services:
   - tag
 
 lint-dockerfiles:
-  stage: pre-build
```
Based on the commit message, I see that the goal of merging the `pre-build` stage into the `build` one was to speed up execution which makes sense.

However, my view for us having both of those stages was that we wanted to avoid  spending computing resources on building the image if there are any issues with the Dockerfile or the Python code and this idea conflicts with the merging of the two stages. So, here I think we should clarify if it's okay to spend time building an image even in the presence of issues on the Dockerfile and/or Python code. Maybe @jsrc27 could help with his view.

Anyways, if the goal is to speed up things, I wanted to ask if you @lucasbernardestoradex were able to understand why those basic jobs in the `pre-build` stage were taking long. Checking the executions on the internal Gitlab I see the stage usually took 4-5 minutes to run but now on `gitlab.com` it's taking 20+ minutes, about the same time the build stage takes. I wonder if the caching on the `tcbuilder-dev-latest` we're employing is actually working...

jsrc27 replied:
> Maybe @jsrc27 could help with his view.

For me I did appreciate having the standalone stage that checks the Dockerfile and runs the Python linter. Let's me get feedback early on any issues with PRs so I can quickly fix things without waiting too long. Now of course this was for the Github pipeline and this change is for Gitlab. Though similar principal would apply.

As @rborn-tx said though on Github usually this initial stage takes maybe several minutes tops, so it's strange that it takes so long on Gitlab.

lucas-p-bernardes replied:
Ok, I agree with you.
Since this is not the initial purpose of the MR, which was the JUnit and XRay integration, I'm droping a12b75e1c4d93f40a0d8e266cd6a47bc76a5ad18 commit by now.

About the point raised that it is taking much longer after migration, I created TCB-690 ticket for this, so we can work in this later avoiding delaying this MR.

lucas-akira replied:
I agree that the `pre-build` issue should be dealt with separately, as it needs a more thorough investigation on what is causing it.

rborn-tx reviewed (commented):
Overall looks good. I've raised a few questions though.

lucas-p-bernardes committed:
tests/integration: Bump BATS versions to get fix in --report-formatter junit

After the implementation of JUnit report, it was noticed some bugs
related to report-formatter finishing after bats execution and invalid
XML characters. In Changelog of BATS-CORE, there are this two fixes
related to this:

- Version 1.9.0: wait for report-formatter to finish before ending bats
execution, to fix empty files with --report-fomatter junit under Docker (#692)
- Version 1.11.0: junit formatter: remove ANSI Codes to avoid invalid XML
character (#886)

Signed-off-by: Lucas Bernardes <lucas.bernardes@toradex.com>

lucas-p-bernardes commented:
Hello @lucas-akira @rborn-tx @jsrc27, can we have another round of review here?

jsrc27 reviewed (approved):
LGTM

rborn-tx reviewed (approved):
LGTM
